src/manage.c

```diff
@@ -192,42 +195,69 @@ void manage_acq_setup()
-
-  for (u8 prn=0; prn<32; prn++) {
-    if (!almanac[prn].valid ||
-        time_quality == TIME_UNKNOWN ||
-        position_quality == POSITION_UNKNOWN) {
-      /* No almanac or position/time information, give it the benefit of the
-       * doubt. */
-      acq_prn_param[prn].score[ACQ_HINT_ALMANAC] = 0;
+    /* Do we have any idea where/when we are?  If not, no score. */
+    /* TODO: Stricter requirement on time and position uncertainty?
+       We ought to keep track of a quantitative uncertainty estimate. */
+    if (time_quality < TIME_GUESS &&
+        position_quality < POSITION_GUESS)
+      return SCORE_COLDSTART;
+
+    float el;
+    double el_d, _, dopp_hint, dopp_uncertainty = DOPP_UNCERT_ALMANAC;
+
+    /* Do we have a suitable ephemeris for this sat?  If so, use
+       that in preference to the almanac. */
+    if (ephemeris_good(&es[prn], t)) {
+      double sat_pos[3], sat_vel[3], el_d;
+      calc_sat_state(&es[prn], t, sat_pos, sat_vel, &_, &_);
+      wgsecef2azel(sat_pos, position_solution.pos_ecef, &_, &el_d);
+      el = (float)(el_d) * R2D;
+      if (el < elevation_mask)
+        return 0;
+      vector_subtract(3, sat_pos, position_solution.pos_ecef, sat_pos);
+      vector_normalize(3, sat_pos);
+      /* sat_pos now holds unit vector from us to satellite */
+      vector_subtract(3, sat_vel, position_solution.vel_ecef, sat_vel);
+      /* sat_vel now holds velocity of sat relative to us */
+      dopp_hint = -GPS_L1_HZ * (vector_dot(3, sat_pos, sat_vel) / GPS_C
+                                + position_solution.clock_bias);
+      /* TODO: Check sign of receiver frequency offset correction */
+      if (time_quality >= TIME_FINE) dopp_uncertainty = DOPP_UNCERT_EPHEM;
+    } else if (almanac[prn].valid) {
+      calc_sat_az_el_almanac(&almanac[prn], t.tow, t.wn-1024,
+                             position_solution.pos_ecef, &_, &el_d);
+      el = (float)(el_d) * R2D;
+      if (el < elevation_mask)
+        return 0;
+      dopp_hint = -calc_sat_doppler_almanac(&almanac[prn], t.tow, t.wn,
+                                            position_solution.pos_ecef);
     } else {
-      calc_sat_az_el_almanac(&almanac[prn], t.tow, t.wn-1024, position_solution.pos_ecef, &az, &el);
-      float dopp = -calc_sat_doppler_almanac(&almanac[prn], t.tow, t.wn,
-                                             position_solution.pos_ecef);
-      acq_prn_param[prn].score[ACQ_HINT_ALMANAC] =
-            el > 0 ? (u16)(SCORE_ALMANAC * 2 * el / M_PI) : 0;
-      acq_prn_param[prn].dopp_hint_low = dopp - ALMANAC_DOPPLER_WINDOW;
-      acq_prn_param[prn].dopp_hint_high = dopp + ALMANAC_DOPPLER_WINDOW;
+      return SCORE_COLDSTART; /* Couldn't determine satellite state. */
     }
-  }
+    /* Return the doppler hints and a score proportional to elevation */
+    *dopp_hint_low = dopp_hint - dopp_uncertainty;
+    *dopp_hint_high = dopp_hint + dopp_uncertainty;
+    return SCORE_COLDSTART + SCORE_WARMSTART * el / 90;
```

Comment:
unit test: I'll make a ticket to add unit tests for all the management functions.  Could be a good one for the "focused tech debt reduction sprint"